theta_dup/include/theta_sketch_dup.h

```diff
@@ -0,0 +1,2209 @@
+  virtual ~compact_theta_sketch_dup_alloc() = default;
+
+  virtual uint32_t get_num_retained() const;
+  virtual uint16_t get_seed_hash() const;
+  virtual bool is_ordered() const;
+  virtual string<A> to_string(bool print_items = false) const;
+  virtual void serialize(std::ostream& os) const;
+  virtual void serialize(datasketches_pb::Update_theta_sketch_dup* pb) const;
+  typedef vector_u8<A> vector_bytes;  // alias for users
+  // header space is reserved, but not initialized
+  virtual vector_bytes serialize(unsigned header_size_bytes = 0) const;
+
+  virtual typename theta_sketch_dup_alloc<A>::const_iterator begin() const;
+  virtual typename theta_sketch_dup_alloc<A>::const_iterator end() const;
+
+  /**
+   * This method deserializes a sketch from a given stream.
+   * @param is input stream
+   * @param seed the seed for the hash function that was used to create the
+   * sketch
+   * @return an instance of a sketch
+   */
+  static compact_theta_sketch_dup_alloc<A> deserialize(
+      std::istream& is, uint64_t seed = DEFAULT_SEED);
+
+  /**
+   * This method deserializes a sketch from protobuf.
+   * @param pb input stream
+   * @param seed the seed for the hash function that was used to create the
+   * sketch
+   * @return an instance of a sketch
+   */
+  static compact_theta_sketch_dup_alloc<A> deserialize(
+      datasketches_pb::Update_theta_sketch_dup& pb,
+      uint64_t seed = DEFAULT_SEED);
+
+  /**
+   * This method deserializes a sketch from a given array of bytes.
+   * @param bytes pointer to the array of bytes
+   * @param size the size of the array
+   * @param seed the seed for the hash function that was used to create the
+   * sketch
+   * @return an instance of the sketch
+   */
+  static compact_theta_sketch_dup_alloc<A> deserialize(
+      const void* bytes, size_t size, uint64_t seed = DEFAULT_SEED);
+
+  /**
+   * @return true if *this equals r
+   */
+  virtual bool isEqual(const compact_theta_sketch_dup_alloc& r) const;
+
+ private:
+  typedef typename std::allocator_traits<A>::template rebind_alloc<uint64_t>
+      AllocU64;
+
+  vector_u64<A> keys_;
+  uint16_t seed_hash_;
+  bool is_ordered_;
+
+  friend theta_sketch_dup_alloc<A>;
+  friend update_theta_sketch_dup_alloc<A>;
+
+  /* TODO: support set operations
+   * friend theta_union_alloc<A>;
+   * friend theta_intersection_alloc<A>;
+   * friend theta_a_not_b_alloc<A>;
+   */
+
+  compact_theta_sketch_dup_alloc(bool is_empty, uint64_t theta,
+                                 vector_u64<A>&& keys, uint16_t seed_hash,
+                                 bool is_ordered);
+  static compact_theta_sketch_dup_alloc<A> internal_deserialize(
+      std::istream& is, uint8_t preamble_longs, uint8_t flags_byte,
+      uint16_t seed_hash);
+  static compact_theta_sketch_dup_alloc<A> internal_deserialize(
+      datasketches_pb::Update_theta_sketch_dup& pb, uint8_t preamble_longs, uint8_t flags_byte,
+      uint16_t seed_hash);
```

Comment:
Declaration: internal_deserialize used by deserialize of the **derived class** using protobuf.